tp_framework/core/discovery.py

```diff
@@ -257,105 +246,367 @@ def discovery(src_dir: Path, l_tp_id: list[int], tp_lib_path: Path, itools: list
+            d_res_tpi[tpi_id] = d_tpi
+        d_res[tp_id]["instances"] = d_res_tpi
+
+    # post-process results and export them
+    d_results = post_process_and_export_results(d_res, build_name, disc_output_dir)
+    d_results["cpg_file"] = str(cpg)
+    logger.info("Discovery for patterns completed.")
+    return d_results
+
 
+def discovery_for_tpi(tpi_instance: Instance, tpi_json_path: Path, cpg: Path, disc_output_dir: Path,
+                      measurement_stop: bool = False, already_executed: dict = {}) -> Dict:
+    msgpre = f"pattern {tpi_instance.pattern_id} instance {tpi_instance.instance_id} - "
+    d_tpi_discovery = {}
+    # execute the discovery rule
+    if not measurement_stop and tpi_instance.discovery_rule:
+        # prepare and execute the discovery rule (if not done yet)
+        dr = (tpi_json_path.parent / tpi_instance.discovery_rule).resolve()
+        logger.info(
+            f"{msgpre}prepare discovery rule {dr}...")
+        d_tpi_discovery["rule_path"] = str(dr)
+        d_tpi_discovery["method"] = tpi_instance.discovery_method
+        d_tpi_discovery["rule_accuracy"] = tpi_instance.discovery_rule_accuracy
+        d_tpi_discovery["rule_hash"] = utils.get_file_hash(dr)
+        if d_tpi_discovery["rule_hash"] not in already_executed:
+            logger.info(
+                f"{msgpre}running discovery rule...")
+            # related to #42
+            pdr = patch_PHP_discovery_rule(dr, tpi_instance.language, output_dir=disc_output_dir)
+            #
+            try:
+                findings = run_and_process_discovery_rule(cpg, pdr, discovery_method=d_tpi_discovery["method"])
+                d_tpi_discovery["results"] = findings
+                d_tpi_discovery["rule_already_executed"] = False
+            except DiscoveryMethodNotSupported as e:
+                d_tpi_discovery["results"] = None
+                already_executed[d_tpi_discovery["rule_hash"]] = None
+                logger.error(
+                    f"{msgpre}Discovery rule failure for this instance: {e}")
+                ## JoernQueryError(e)
+            ## JoernQueryParsingResultError(e)
+            already_executed[d_tpi_discovery["rule_hash"]] = findings
+            logger.info(
+                f"{msgpre} discovery rule executed.")
+
+        else:
+            logger.info(
+                f"{msgpre}discovery rule {dr} was already run. We will use the same result...")
+            d_tpi_discovery["rule_already_executed"] = True
+            d_tpi_discovery["results"] = already_executed[d_tpi_discovery["rule_hash"]]
+    else:
+        # no rule to execute
+        d_tpi_discovery["rule_path"] = None
+        d_tpi_discovery["method"] = None
+        d_tpi_discovery["rule_name"] = None
+        d_tpi_discovery["rule_accuracy"] = None
+        d_tpi_discovery["rule_hash"] = None
+        d_tpi_discovery["rule_name"] = None
+        d_tpi_discovery["results"] = None

```

Comment:
```suggestion
        # no rule to execute
        d_tpi_discovery["rule_already_executed"] = False
```
With the changes from above, we could get rid of all these lines and replace it just with the `d_tpi_discovery["rule_already_executed"] = False` line